torch_xla/csrc/init_python_bindings.cpp

```diff
@@ -103,59 +103,78 @@ class PythonScope : public Scope {
+    using ParameterTypes =
+        typename c10::guts::infer_function_traits<Func>::type::parameter_types;
+    PythonFunctionBinder<FunctionKind::kStatic, ParameterTypes, Extra...>::Bind(
         *this, name, std::forward<Func>(f), extra...);
     return *this;
   }
 
  private:
-  template <typename F, typename RetType, typename ParameterTypes,
-            typename... Extra>
-  struct PythonFunctionBinder;
+  // The kind of python function.
+  enum class FunctionKind {
+    kInit,     // __init__
+    kRegular,  // a regular (non-static) function
+    kStatic,   // a static function
+  };
 
-  template <typename F, typename RetType, typename... Args, typename... Extra>
-  struct PythonFunctionBinder<
-      F, RetType, c10::guts::typelist::typelist<Args...>, Extra...> {
-    // Binds an __init__ function in the given scope.
-    static void BindInit(Scope& scope, F&& f, const Extra&... extra) {
-      scope.def(py::init(
-          [f = std::move(f)](Args... args) -> RetType {
-            torch::PyWarningHandler handler;
-            return f(args...);
-          },
-          extra...));
-    }
+  template <FunctionKind kind, typename ParameterTypes, typename... Extra>
+  struct PythonFunctionBinder;
 
+  template <FunctionKind kind, typename... Args, typename... Extra>
+  struct PythonFunctionBinder<kind, c10::guts::typelist::typelist<Args...>,
+                              Extra...> {
     // Binds a python function with the given name in the given scope.
+    // When scope is kInit, the name is ignored.
+    template <typename F>
     static void Bind(Scope& scope, const char* const name, F&& f,
                      const Extra&... extra) {
-      scope.def(
-          name,
-          [f = std::move(f)](Args... args) -> RetType {
-            // RAII for emitting Python warnings.
-            //
-            // This turns messages passed to `TORCH_WARN()` in `f` into Python
-            // warnings.
-            torch::PyWarningHandler handler;
-            return f(args...);
-          },
-          extra...);
+      using RetType =
+          typename c10::guts::infer_function_traits<F>::type::return_type;
+
+      if constexpr (kind == FunctionKind::kInit) {
+        scope.def(py::init(
+            [f = std::move(f)](Args... args) -> RetType {
+              torch::PyWarningHandler handler;
+              return f(args...);
+            },
+            extra...));
+      } else {
+        auto lambda = [f = std::move(f)](Args... args) -> RetType {
+          // RAII for emitting Python warnings.
+          //
+          // This turns messages passed to `TORCH_WARN()` in `f` into Python
+          // warnings.
+          torch::PyWarningHandler handler;
+          return f(args...);
+        };
+        if constexpr (kind == FunctionKind::kStatic) {
+          scope.def_static(name, std::move(lambda), extra...);
+        } else {
+          scope.def(name, std::move(lambda), extra...);
+        }
+      }
```

Comment:
What do you think of unconditionally creating a `lambda` variable first (since all execution paths require it), and then call `scope.def*()` as needed?